sstables/sstables.cc

```diff
@@ -1275,6 +1275,17 @@ future<> sstable::open_data() {
         if (_shards.empty()) {
             _shards = compute_shards_for_this_sstable();
         }
+        // Sharding information uses a lot of memory and once we're doing with this computation we will no longer use it.
+        // Destruction of large containers can cause stalls, so protect against it.
+        // This is not perf critical and a thread looks better.
+        return seastar::async([this] {
+            auto* sm = _components->scylla_metadata->data.get<scylla_metadata_type::Sharding, sharding_metadata>();
+            while (sm && !sm->token_ranges.elements.empty()) {
```

Comment:
Can be simplified to

```c++
return do_until([c = &sm->token_ranges.elements] { return c->empty(); },
    c->pop_back();
}).then([this] {
    sm->_components.scylla_metadata = {}
});
```

But _components is a foreign_ptr, please verify this is never called after we shared it.

Reply:
Indeed `_components` is a foreign_ptr, but it seems to be only invoked remotely by `get_open_info` that does this:

```
future<foreign_sstable_open_info> sstable::get_open_info() & {
    return _components.copy().then([this] (auto c) mutable {                                                                                                               
        return foreign_sstable_open_info{std::move(c), this->get_shards_for_this_sstable(), _data_file.dup(), _index_file.dup(),                                           
            _generation, _version, _format};                                                                                                                               
    });
}  
```

First weird thing here is that it actually creates a `.copy()`. What is the point of having shared components if we create a copy ?

But one way or another, the sharding information is already copied post-parsed. So we never access that particular container cross-CPU even if we do access the `_components` pointer

Reply:
It copy()s the pointer, not the value.

Reply:
Yes, cross-cpu loads don't use it:

```c++
future<> sstable::load(sstables::foreign_sstable_open_info info) {
    return read_toc().then([this, info = std::move(info)] () mutable {
        _components = std::move(info.components);
        _data_file = make_checked_file(_read_error_handler, info.data.to_file());
        _index_file = make_checked_file(_read_error_handler, info.index.to_file());
        _shards = std::move(info.owners);
        validate_min_max_metadata();
        validate_max_local_deletion_time();
        validate_partitioner();
        return update_info_for_opened_data();
    });
}
```

Reply:
Yes, my understanding from the code above is that we're using the post-parsed information recorded in `owners`

Reply:
It is somewhat fragile since we keep the object in memory. It would be better to read that as a temporary and never store it in the sstable, but we can't, because it's part of a larger object that does need to be kept in the sstable.